<h1 style='text-align: center;'>UK ROAD ACCIDENT DATA ANALYSIS</h1>
<br>
<h2 style='text-align: center;'>JAY CEE I. MACARAIG</h2>
<h3 style='text-align: center;'>BSIT - 3B</h3>
<hr>

<h2>IMPORTING IMPORTANT LIBRARIES</h2>

In [1]:
from scipy.stats import f_oneway
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import warnings

warnings.filterwarnings('ignore')

In [2]:
data = pd.read_csv('datasets\\uk_road_accident.csv')

In [3]:
data

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<hr><br>
<h2>CLEANING DATA TYPES AND FILLING UP NULL VALUES</h2>

In [4]:
data.columns

Index(['Index', 'Accident_Severity', 'Accident Date', 'Latitude',
       'Light_Conditions', 'District Area', 'Longitude',
       'Number_of_Casualties', 'Number_of_Vehicles', 'Road_Surface_Conditions',
       'Road_Type', 'Urban_or_Rural_Area', 'Weather_Conditions',
       'Vehicle_Type'],
      dtype='object')

In [5]:
data['Index'] = data['Index'].astype('category')
data['Accident_Severity'] = data['Accident_Severity'].astype('category')
data['Accident Date'] = data['Accident Date'].astype('category')
data['Light_Conditions'] = data['Light_Conditions'].astype('category')
data['District Area'] = data['District Area'].astype('category')
data['Road_Surface_Conditions'] = data['Road_Surface_Conditions'].astype('category')
data['Road_Type'] = data['Road_Type'].astype('category')
data['Urban_or_Rural_Area'] = data['Urban_or_Rural_Area'].astype('category')
data['Weather_Conditions'] = data['Weather_Conditions'].astype('category')
data['Vehicle_Type'] = data['Vehicle_Type'].astype('category')

In [6]:
data.dtypes

Index                      category
Accident_Severity          category
Accident Date              category
Latitude                    float64
Light_Conditions           category
District Area              category
Longitude                   float64
Number_of_Casualties          int64
Number_of_Vehicles            int64
Road_Surface_Conditions    category
Road_Type                  category
Urban_or_Rural_Area        category
Weather_Conditions         category
Vehicle_Type               category
dtype: object

In [7]:
data

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [8]:
data.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [9]:
data['Latitude'] = data['Latitude'].fillna(data['Latitude'].mean())
data['Longitude'] = data['Longitude'].fillna(data['Longitude'].mean())
data['Road_Surface_Conditions'] = data['Road_Surface_Conditions'].fillna(data['Road_Surface_Conditions'].mode()[0])
data['Road_Type'] = data['Road_Type'].fillna(data['Road_Type'].mode()[0])
data['Urban_or_Rural_Area'] = data['Urban_or_Rural_Area'].fillna(data['Urban_or_Rural_Area'].mode()[0])
data['Weather_Conditions'] = data['Weather_Conditions'].fillna(data['Weather_Conditions'].mode()[0])

In [10]:
data.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<br>
<br>
<h2 style='text-align: center;'>DATA EXPLORATORY ANALYSIS</h2>
<hr>

<br>
<br>
<h2>QUESTION 1: WHICH DISTRICT AREA HOLDS THE MOST FATAL ACCIDENTS?</h2>
<hr>

In [93]:
fatal_accidents = data[data['Accident_Severity'] == 'Fatal']

fatal_accidents['District Area'].value_counts().sort_values(ascending=False)

District Area
Birmingham                  105
Leeds                        93
Highland                     88
East Riding of Yorkshire     85
Bradford                     71
                           ... 
Orkney Islands                2
Havant                        2
Stevenage                     1
Clackmannanshire              1
Harlow                        1
Name: count, Length: 422, dtype: int64

<br>
<h4>The data analysis shown above says that <strong>BIRMINGHAM</strong> holds the most fatal accidents in UK.</h4>
<p><i>INSIGHT:</i> Birmingham has 105 number of fatal accidents, making it.</p>
<hr>
<br>

<br>
<br>
<h2>QUESTION 2: WHERE DO MORE SEVERE ACCIDENTS HAPPEN? IS IT IN URBAN OR RURAL AREAS?</h2>
<hr>

In [12]:
severe_accidents = data[data['Accident_Severity'] == 'Serious']

severe_accidents['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          50904
Rural          37312
Unallocated        1
Name: count, dtype: int64

<br>
<h4>The data analysis shown above says that <strong>URBAN AREA</strong> has more severe accidents in the UK.</h4>
<p><i>INSIGHT:</i> Urban areas has 20k more accidents than Rural areas.</p>
<hr>
<br>

<br>
<br>
<h2>QUESTION 3: WHICH VEHICLE TYPE HAS THE HIGHEST NUMBER OF CASUALTIES?</h2>
<hr>

In [13]:
data.groupby('Vehicle_Type')['Number_of_Casualties'].sum().sort_values(ascending=False)

Vehicle_Type
Car                                      676692
Van / Goods 3.5 tonnes mgw or under       46271
Bus or coach (17 or more pass seats)      34915
Motorcycle over 500cc                     34879
Goods 7.5 tonnes mgw and over             23397
Motorcycle 125cc and under                20348
Taxi/Private hire car                     18195
Motorcycle over 125cc and up to 500cc     10286
Motorcycle 50cc and under                 10167
Goods over 3.5t. and under 7.5t            8308
Other vehicle                              7554
Minibus (8 - 16 passenger seats)           2659
Agricultural vehicle                       2613
Pedal cycle                                 270
Data missing or out of range                  9
Ridden horse                                  5
Name: Number_of_Casualties, dtype: int64

<br>
<h4>The data analysis showsn above says that <strong>CAR TYPE VEHICLE</strong> has the highest number of casualties when it comes to road accidents in the UK.</h4>
<p><i>INSIGHT:</i> Car being the vehicle type with the highest casualties is maybe due to its being the most common owned vehicle in the nation.</p>
<hr>
<br>

<br>
<br>
<h2>QUESTION 4: WHAT WEATHER CONDITIONS HOST THE MOST ACCIDENTS IN A SINGLE CARRIAGEWAY TYPE OF ROAD?</h2>
<hr>

In [14]:
single_carriageway = data[data['Road_Type'] == 'Single carriageway']

single_carriageway['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       403388
Raining no high winds     59011
Other                     13288
Raining + high winds       6759
Fine + high winds          6228
Snowing no high winds      4781
Fog or mist                2562
Snowing + high winds        646
Name: count, dtype: int64

<br>
<h4>The data analysis shown above says that the weather condition with the highest accidents occuring in a single carriageway type of road is usually <strong>Fine no high winds</strong></h4>
<p><i>INSIGHT:</i> The data above shows that accidents occur usually in a normal day in the UK.</p>
<hr>
<br>

<br>
<br>
<h2>QUESTION 5: IN WHAT TIME OF THE DAY OR LIGHT CONDITIONS MOST ACCIDENTS HAPPEN?</h2>
<hr>

In [91]:
data['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<br>
<h4>The data analysis shown above says that its usually in <strong>Daylight</strong> that most accidents in the UK happends.</h4>
<p><i>INSIGHT:</i> The data above shows that accidents occur usually in a morning in the UK, with the value of 484,880 accidents in a Daylight setting.</p>
<hr>
<br>

<br>
<br>
<h2>QUESTION 6: IN WHAT TYPE OF ROAD DOES MOST ACCIDENTS HAPPEN?</h2>
<hr>

In [16]:
data['Road_Type'].value_counts()

Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

<br>
<h4>The data analysis shown above says that its usually in <strong>Single carriageway</strong> that most accidents in the UK happends.</h4>
<p><i>INSIGHT:</i> The data above shows that accidents occur usually in single carriageway in the UK, with the value of 496,663 accidents.</p>
<hr>
<br>

<br>
<br>
<h2>QUESTION 7: IN WHAT DISTRICT AREA HAD THE ACCIDENT WITH THE MOST CASUALTIES THAT IS EVER RECORDED?</h2>
<hr>

In [92]:
highest_casualties = data['Number_of_Casualties'].max()

print(f'THE HIGHEST CASUALTIES IN ONE ACCIDENT EVER RECORDED IS {highest_casualties}')

data[data['Number_of_Casualties'] == highest_casualties]

THE HIGHEST CASUALTIES IN ONE ACCIDENT EVER RECORDED IS 68


Index Accident_Severity Accident Date   Latitude  \
117980  200743N002017             Fatal      3/1/2019  51.497547   

             Light_Conditions District Area  Longitude  Number_of_Casualties  \
117980  Darkness - lights lit   South Bucks  -0.496697                    68   

        Number_of_Vehicles Road_Surface_Conditions  Road_Type  \
117980                   1             Wet or damp  Slip road   

       Urban_or_Rural_Area     Weather_Conditions Vehicle_Type  
117980               Rural  Raining no high winds          Car

<br>
<h4>The data analysis shown above says that the accident with the most casualties ever recorded in the dataset is in <strong>South Bucks</strong> district.</h4>
<p><i>INSIGHT:</i> The data above shows that number of casualties is 68 and it happened on January 3rd, 2019</p>
<hr>
<br>

<br>
<br>
<h2>QUESTION 8: WHAT'S THE AVERAGE NUMBER OF CASUALTIES BY THE SEVERITY OF ACCIDENTS?</h2>
<hr>

In [18]:
np.round(data.groupby('Accident_Severity')['Number_of_Casualties'].mean(),2)

Accident_Severity
Fatal      1.90
Serious    1.47
Slight     1.33
Name: Number_of_Casualties, dtype: float64

<br>
<h4>When it comes to Fatal Accidents, the average number of casualties is <strong>1.90</strong>.</h4>
<h4>While, when it comes to Serious Accidents, it is <strong>1.47</strong>.</h4>
<h4>Lastly, when it comes to Slight Accidents, it is <strong>1.33</strong>.</h4>
<p><i>INSIGHT:</i> The data above shows that the average number of casualties, be it in every severity of accidents does not exceeds to more than 2. </p>
<hr>
<br>

<br>
<br>
<h2>QUESTION 9: WHAT IS THE MOST FREQUENT ROAD SURFACE CONDITION THAT IS FOUND IN URBAN AREA WHEN AN ACCIDENT HAPPENS?</h2>
<hr>

In [24]:
urban_area = data[data['Urban_or_Rural_Area'] == 'Urban']

urban_area['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     303397
Wet or damp             107698
Frost or ice              7564
Snow                      2788
Flood over 3cm. deep       231
Name: count, dtype: int64

<br>
<h4>The data shown above says that the surface of the road is usually <strong>Dry</strong> when an accident happens.</h4>
<p><i>INSIGHT:</i> Dry and wet/damp are the common road surface condition when an accident happens in the UK. </p>
<hr>
<br>

<br>
<br>
<h2>QUESTION 10: WHAT DATE HAS THE HIGHEST ROAD ACCIDENTS EVER RECORDED? HOW MANY CASUALTIES WERE AFFECTED THAT DATE? </h2>
<hr>

In [112]:
most_frequent_date = data['Accident Date'].mode()[0]
total_casualties = data[data['Accident Date'] == most_frequent_date]['Number_of_Casualties'].sum()

print(f'THE DATE WITH THE HIGHEST ROAD ACCIDENT IS {most_frequent_date}')
print(f'WHILE THE TOTAL CASUALTIES DURING THAT DAY IS {total_casualties}')

data[data['Accident Date'] == most_frequent_date]

THE DATE WITH THE HIGHEST ROAD ACCIDENT IS 30-11-2019
WHILE THE TOTAL CASUALTIES DURING THAT DAY IS 938


Index Accident_Severity Accident Date   Latitude  \
627     200701BS70671            Slight    30-11-2019  51.495858   
629     200701BS70673            Slight    30-11-2019  51.523668   
639     200701BS70685            Slight    30-11-2019  51.490534   
647     200701BS70694            Slight    30-11-2019  51.505260   
674     200701BS70725           Serious    30-11-2019  51.484887   
...               ...               ...           ...        ...   
181141  200797UA70811           Serious    30-11-2019  55.642006   
181945       2.01E+12           Serious    30-11-2019  55.050682   
181946       2.01E+12            Slight    30-11-2019  55.031332   
181949       2.01E+12           Serious    30-11-2019  55.068908   
182100       2.01E+12            Slight    30-11-2019  55.113411   

              Light_Conditions           District Area  Longitude  \
627      Darkness - lights lit  Kensington and Chelsea  -0.192343   
629      Darkness - lights lit  Kensington and Chelsea  -0.217334   
639                   Daylight  Kensington and Chelsea  -0.191113   
647      Darkness - lights lit  Kensington and Chelsea  -0.213153   
674      Darkness - lights lit  Kensington and Chelsea  -0.174917   
...                        ...                     ...        ...   
181141                Daylight          North Ayrshire  -4.683832   
181945                Daylight   Dumfries and Galloway  -3.776910   
181946  Darkness - no lighting   Dumfries and Galloway  -3.617705   
181949                Daylight   Dumfries and Galloway  -3.610294   
182100                Daylight   Dumfries and Galloway  -3.405752   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
627                        1                   2             Wet or damp   
629                        3                   3             Wet or damp   
639                        1                   1                     Dry   
647                        1                   1             Wet or damp   
674                        1                   2             Wet or damp   
...                      ...                 ...                     ...   
181141                     1                   1             Wet or damp   
181945                     1                   1             Wet or damp   
181946                     1                   1             Wet or damp   
181949                     1                   1             Wet or damp   
182100                     1                   2             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
627     Single carriageway               Urban  Raining no high winds   
629     Single carriageway               Urban  Raining no high winds   
639     Single carriageway               Urban     Fine no high winds   
647       Dual carriageway               Urban  Raining no high winds   
674     Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
181141  Single carriageway               Rural     Fine no high winds   
181945  Single carriageway               Rural     Fine no high winds   
181946  Single carriageway               Rural  Raining no high winds   
181949      One way street               Urban  Raining no high winds   
182100  Single carriageway               Rural     Fine no high winds   

                               Vehicle_Type  
627                                     Car  
629           Goods 7.5 tonnes mgw and over  
639     Van / Goods 3.5 tonnes mgw or under  
647                   Motorcycle over 500cc  
674              Motorcycle 125cc and under  
...                                     ...  
181141                                  Car  
181945                                  Car  
181946                                  Car  
181949                                  Car  
182100                                  Car  

[704 rows x 14 columns]

<br>
<h4>The data above shows that the date with the highest accidents is on <strong>November 30th, 2019</strong>.</h4>
<h4>While, the total casualties falls into <strong>938</strong> during that day.</h4>
<p><i>INSIGHT:</i> A total of 704 accidents during that day was recorded, all happening across the nation of UK.</p>
<hr>
<br>

<br>
<br>
<h2>QUESTION 11: DOES THE GEOGRAPHICAL AREA (URBAN & RURAL) HAS A SIGNIFICANT EFFECT TO THE NUMBER OF CASUALTIES PER ACCIDENT? </h2>
<hr>

In [122]:
urban_casualties = data[data['Urban_or_Rural_Area'] == 'Urban']['Number_of_Casualties']
rural_casualties = data[data['Urban_or_Rural_Area'] == 'Rural']['Number_of_Casualties']


stats, p_value = f_oneway(urban_casualties, rural_casualties)

print(f'The probability value is {p_value}')
if p_value >= 0.5:
    print(f'Hence, it does have a correlation.')
else:
    print(f'Hence, it does not have a correlation.')

# if probability value >= 0.5 = with correlation
# else if probability value < 0.5 = no correlation

The probability value is 0.0
Hence, it does not have a correlation.


<br>
<h4>The analysis above shows that the Geograhpical area <strong>do not have a correlation</strong> to the number of casualties per accident.</h4>
<p><i>INSIGHT:</i> The probability that the two variable has a correlation is only 0.0</p>
<hr>
<br>

<br>
<br>
<h2>QUESTION 12: DOES THE NUMBER OF VEHICLES PER ACCIDENT HAVE A CORRELATION TO THE NUMBER OF CASUALTIES? </h2>
<hr>

In [126]:
no_of_vehicles = data['Number_of_Vehicles']
no_of_casualties = data['Number_of_Casualties']

p_value = no_of_vehicles.corr(no_of_casualties)

print(f'The probability value is {p_value}')
if p_value >= 0.5:
    print(f'Hence, it does have a correlation.')
else:
    print(f'Hence, it does not have a correlation.')

# if probability value >= 0.5 = with correlation
# else if probability value < 0.5 = no correlation

The probability value is 0.22888886126927557
Hence, it does not have a correlation.


<br>
<h4>The analysis above shows that the NUmver <strong>do not have a correlation</strong> to the number of casualties per accident.</h4>
<p><i>INSIGHT:</i> The probability that the two variable has a correlation is only 0.0</p>
<hr>
<br>

In [124]:
data.dtypes

Index                      category
Accident_Severity          category
Accident Date              category
Latitude                    float64
Light_Conditions           category
District Area              category
Longitude                   float64
Number_of_Casualties          int64
Number_of_Vehicles            int64
Road_Surface_Conditions    category
Road_Type                  category
Urban_or_Rural_Area        category
Weather_Conditions         category
Vehicle_Type               category
dtype: object